File: Problems/1345. (H) Jump Game IV/Java/solution.java
class Solution {
  public int minJumps(int[] arr) {
    // bfs, use dictionary mapping value to indices

    Map<Integer,List<Integer>> v2i = new HashMap();
    for (int i = 0; i < arr.length; i++) {
      if (!v2i.containsKey(arr[i]))
        v2i.put(arr[i], new ArrayList());
      v2i.get(arr[i]).add(i);
    }

    Deque<Pair<Integer,Integer>> nodes = new ArrayDeque();
    Set<Integer> visited = new HashSet();
    nodes.add(new Pair<Integer,Integer>(0,0));
    while (nodes.size() != 0) {
      Pair<Integer,Integer> cur = nodes.removeFirst();
      int idx = cur.getKey();
      int jmp = cur.getValue();
      if (visited.contains(idx))
        continue;
      visited.add(idx);
      if (idx == arr.length-1)
        return jmp;
      if (v2i.containsKey(arr[idx])) {
        // Java for-each
        for (int i: v2i.get(arr[idx]))
          nodes.add(new Pair<Integer,Integer>(i,jmp+1));
        // no longer need to consider jumps between indices with value arr[idx]
        v2i.remove(arr[idx]);
      }
      if (idx != 0)
        nodes.add(new Pair<Integer,Integer>(idx-1,jmp+1));
      nodes.add(new Pair<Integer,Integer>(idx+1,jmp+1));
    }
    // this line should never execute given problem description
    return -1;
  }
}

File: Problems/1345. (H) Jump Game IV/Python3/solution.py
class Solution:
  def minJumps(self, arr: List[int]) -> int:
    # bfs, use dict mapping value to indices

    v2i = {}
    for i, n in enumerate(arr):
      if n not in v2i:
        v2i[n] = []
      v2i[n].append(i)
    
    nodes = deque()
    visited = set()
    nodes.append((0, 0))
    while nodes:
      idx, jmp = nodes.popleft()
      if idx in visited:
        continue
      visited.add(idx)
      if idx == len(arr)-1:
        return jmp
      if arr[idx] in v2i:
        # consider furthest index first
        for i in reversed(v2i[arr[idx]]):
          nodes.append((i, jmp+1))
        # no longer need to consider jumps between indices with value arr[idx]
        del v2i[arr[idx]]
      if idx:
        nodes.append((idx-1, jmp+1))
      nodes.append((idx+1, jmp+1))
    
    # this line should never execute given problem description
    return -1

